### PR TITLE
Assignment 5

### PR DESCRIPTION
**Pull request by MarharytaFilipovych**

### CODE REVIEW

File: pa5.sql [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/*Create two different views whith the same logical load: select for Europe and for North America 
+  trails ordered by their popularity which values number of posted 
+reviews. Result set should include: trail name, image, coordinates, description, 
+  first sentenses of two most recent reviews (comma-separated).*/
+
+-- SET @na_poly = 'POLYGON((90 -168.75, 90 -10, 78.13 -10, 57.5 -37.5, 15 -30, 15 -75, 1.25 -105, 51 -180, 60 -180, 60 -168.75, 90 -168.75))';
+-- SET @eur_poly = 'POLYGON((90 -10, 90 77.5, 42.5 48.8, 42.5 30, 40.79 28.81, 41 29, 40.55 27.31, 40.40 26.75, 40.05 26.36, 39.17 25.19, 35.46 27.91, 33 27.5, 38 10, 35.42 -10, 28.25 -13, 15 -30, 57.5 -37.5, 78.13 -10, 90 -10))';
+
+CREATE VIEW popular_trails_from_europe AS
```
Comment (Ivashchyk) [inline pa5.sql]:
```suggestion
CREATE VIEW popular_trails_from_europe_view AS 
```

File: pa5.sql [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+         INNER JOIN review r ON r.trail_id = t.id
+         INNER JOIN ( SELECT  *, ROW_NUMBER() OVER (PARTITION BY r.trail_id ORDER BY r.review_date DESC) AS number FROM review r) AS most_recent_reviews ON most_recent_reviews.id = r.id
+WHERE most_recent_reviews.number <= 2 AND ST_CONTAINS(ST_GEOMFROMTEXT('POLYGON((90 -168.75, 90 -10, 78.13 -10, 57.5 -37.5, 15 -30, 15 -75, 1.25 -105, 51 -180, 60 -180, 60 -168.75, 90 -168.75))'), POINT(l.latitude, l.longitude))
+GROUP BY
+    t.name, t.image, coordinates, t.description
+ORDER BY
+    COUNT(t.id) DESC;
+
+
+SELECT * FROM popular_trails_from_europe;
```
Comment (Ivashchyk) [inline pa5.sql]: Do not use * in production code

File: pa4_procedures.sql [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+BEGIN SELECT t.name FROM trail t 
+    INNER JOIN review r ON r.trail_id = t.id
+    INNER JOIN user u ON r.user_id = u.id
+    WHERE u.name = person;
+END //
+DELIMITER ;
+-- update the description of a trail given its name. I can call this stored procedure by providing the name of the trail and the new description. 
+-- After execution, I can retrieve the updated description from the new_description parameter.
+DELIMITER //
+CREATE PROCEDURE update_trail_description_sproc(IN trail_name VARCHAR(50), INOUT new_description TEXT)
```
Comment (Ivashchyk) [inline pa4_procedures.sql:25]: INOUT param is not used as INOUT, just as IN kind

File: pa4_procedures.sql [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    WHERE u.name = user_name
+    LIMIT 1;
+END//
+DELIMITER ;
+-- This stored procedure checks if a user with a given email exists in the database. It returns 1 for true, 0 for false, indicating whether the user exists or not.
+
+DELIMITER //
+CREATE PROCEDURE check_if_user_exists_by_their_email_sproc(IN email VARCHAR(50), OUT user_exists TINYINT)
+    BEGIN
+        SELECT COUNT(*) INTO user_exists FROM user u
```
Comment (Ivashchyk) [inline pa4_procedures.sql]:
```suggestion
        SELECT COUNT(1) INTO user_exists FROM user u 
```

File: pa4_procedures.sql [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    DECLARE trail_exists INT;
+    
+    START TRANSACTION ;
+
+    IF LENGTH(new_description) = 0 OR LENGTH(new_description)> 255 THEN
+        ROLLBACK;
+        SELECT 'Rollback: New description is empty or exceeds the maximum length of 255 characters.' AS message;
+    ELSE
+
+        SELECT COUNT(*) INTO trail_exists FROM trail t WHERE t.name = trail_name;
```
Comment (Ivashchyk) [inline pa4_procedures.sql]:
```suggestion
        SELECT COUNT(1) INTO trail_exists FROM trail t WHERE t.name = trail_name;
```

File: pa5.sql [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+         INNER JOIN review r ON r.trail_id = t.id
+         INNER JOIN ( SELECT  *, ROW_NUMBER() OVER (PARTITION BY r.trail_id ORDER BY r.review_date DESC) AS number FROM review r) AS most_recent_reviews ON most_recent_reviews.id = r.id
+WHERE most_recent_reviews.number <= 2 AND ST_CONTAINS(ST_GEOMFROMTEXT('POLYGON((90 -10, 90 77.5, 42.5 48.8, 42.5 30, 40.79 28.81, 41 29, 40.55 27.31, 40.40 26.75, 40.05 26.36, 39.17 25.19, 35.46 27.91, 33 27.5, 38 10, 35.42 -10, 28.25 -13, 15 -30, 57.5 -37.5, 78.13 -10, 90 -10))'), POINT(l.latitude, l.longitude))
+GROUP BY
+    t.name, t.image, coordinates, t.description
+ORDER BY
+    COUNT(t.id) DESC;
+
+
+CREATE VIEW popular_trails_from_north_america AS
```
Comment (Ivashchyk) [inline pa5.sql]:
```suggestion
CREATE VIEW popular_trails_from_north_america_view AS 
```

File: pa5.sql [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+       t.image,
+       CONCAT(l.longitude, '; ', l.latitude ) AS coordinates,
+       t.description,
+       GROUP_CONCAT(
+               SUBSTRING_INDEX(most_recent_reviews.content, '.', 1) SEPARATOR ', '
+       ) AS recent_reviews
+FROM trail t
+         INNER JOIN location l ON t.location_id = l.id
+         INNER JOIN review r ON r.trail_id = t.id
+         INNER JOIN ( SELECT  *, ROW_NUMBER() OVER (PARTITION BY r.trail_id ORDER BY r.review_date DESC) AS number FROM review r) AS most_recent_reviews ON most_recent_reviews.id = r.id
```
Comment (Ivashchyk) [inline pa5.sql:46]: Do not use * in production code

File: pa5.sql [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+WHERE most_recent_reviews.number <= 2 AND ST_CONTAINS(ST_GEOMFROMTEXT('POLYGON((90 -168.75, 90 -10, 78.13 -10, 57.5 -37.5, 15 -30, 15 -75, 1.25 -105, 51 -180, 60 -180, 60 -168.75, 90 -168.75))'), POINT(l.latitude, l.longitude))
+GROUP BY
+    t.name, t.image, coordinates, t.description
+ORDER BY
+    COUNT(t.id) DESC;
+
+
+SELECT * FROM popular_trails_from_europe;
+
+SELECT * FROM popular_trails_from_north_america;
```
Comment (Ivashchyk) [inline pa5.sql]: ditto

**Review by Ivashchyk — CHANGES_REQUESTED** [review]
1. See comments
2. Why does your PR contains so many files and most of them with commented code? Seems to be dangerous, please review your changes one more time.

Overall, very good performance 💪

**Comment by MarharytaFilipovych** [comment]
Those strange files have appeared on their own, by default, when I created a program to populate trail_rating table. In order to create that program, I had to download some updates to use mysqlconnection in program. It is not the first time. When I set up this connection, lots of strange files usually appear. I always try to add them to .gitignore, but in this case I was not able to add all of them. @Ivashchyk

**Pull request merged**